be/src/storage/lake/compaction_policy.cpp

```diff
@@ -37,6 +39,93 @@ class BaseAndCumulativeCompactionPolicy : public CompactionPolicy {
     TabletMetadataPtr _tablet_metadata;
 };
 
+struct RowsetStat {
+    size_t num_rows = 0;
+    size_t num_dels = 0;
+    size_t bytes = 0;
+};
+
+class PrimaryCompactionPolicy : public CompactionPolicy {
+public:
+    using RowsetCandidate = std::pair<RowsetMetadataPtr, RowsetStat>;
+    explicit PrimaryCompactionPolicy(TabletPtr tablet) : _tablet(std::move(tablet)) {}
+    ~PrimaryCompactionPolicy() override = default;
+
+    StatusOr<std::vector<RowsetPtr>> pick_rowsets(int64_t version) override;
+
+private:
+    double calc_compaction_score(const RowsetStat& stats);
+    static const size_t compaction_result_bytes_threashold = 1000000000;
+    static const size_t compaction_result_rows_threashold = 10000000;
+
+private:
+    TabletPtr _tablet;
+};
+
+StatusOr<std::vector<RowsetPtr>> PrimaryCompactionPolicy::pick_rowsets(int64_t version) {
+    ASSIGN_OR_RETURN(auto tablet_metadata, _tablet->get_metadata(version));
+
+    std::vector<RowsetPtr> input_rowsets;
+    // The goal of lake primary table compaction:
+    // 1. clean up deleted bytes.
+    // 2. merge small rowsets to bigger rowset.
+    // so we pick rowset to compact by this logic:
+    // First, pick out rowset with more deleted bytes.
+    // Second, pick out rowset with less bytes.
+    struct RowsetCompare {
+        static double calc_del_bytes(const RowsetCandidate& rc) {
+            return (double)rc.second.bytes * (double)rc.second.num_dels / (double)rc.second.num_rows;
+        }
+        bool operator()(const RowsetCandidate& a, const RowsetCandidate& b) const {
+            double delete_bytes_a = calc_del_bytes(a);
+            double delete_bytes_b = calc_del_bytes(b);
+            if (delete_bytes_a < delete_bytes_b) {
+                return true;
+            } else if (delete_bytes_a > delete_bytes_b) {
+                return false;
+            } else {
+                // may happen when deleted rows is zero
+                return a.second.bytes > b.second.bytes;
+            }
+        }
+    };
```

Comment:
Nit: how about define `RowsetCandidate` as a new class and make a custom '<' operator of it?

Reply:
Done